File: 2024/6/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split(""));

class Guard {
  constructor() {
    this.dir = [0, -1];
  }

  setPos(pos) {
    this.pos = pos;
  }

  turn() {
    this.dir = [-this.dir[1], this.dir[0]];
  }

  patrol(patrolArea = input) {
    this.dir = [0, -1];
    let [r, c] = this.pos;
    const patrolPath = new Set();
    const patrolPathWithDirection = new Set();
    while (
      r >= 0 &&
      r < patrolArea.length &&
      c >= 0 &&
      c < patrolArea[r].length
    ) {
      patrolPath.add(`${r},${c}`);
      while (
        r + this.dir[1] >= 0 &&
        r + this.dir[1] < patrolArea.length &&
        c + this.dir[0] >= 0 &&
        c + this.dir[0] < patrolArea[r].length &&
        patrolArea[r + this.dir[1]][c + this.dir[0]] === "#"
      ) {
        this.turn();
      }
      if (patrolPathWithDirection.has(`${r},${c},${this.dir.join(",")}`))
        return null;
      patrolPathWithDirection.add(`${r},${c},${this.dir.join(",")}`);
      r += this.dir[1];
      c += this.dir[0];
    }
    return patrolPath;
  }
}

const guard = new Guard();
const obstructionRows = new Map();
const obstructionCols = new Map();

for (let r = 0; r < input.length; r++) {
  for (let c = 0; c < input[r].length; c++) {
    if (input[r][c] === "^") {
      guard.setPos([r, c]);
      break;
    }
  }
}

const patrolPath = guard.patrol();
console.log(patrolPath.size);

let part2 = 0;
for (const coord of patrolPath) {
  const [r, c] = coord.split(",").map(Number);
  input[r][c] = "#";
  const tmpPath = guard.patrol(input);
  if (!tmpPath) {
    part2++;
  }
  input[r][c] = ".";
}
console.log(part2);

File: 2024/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split((string(data)), "\n")
	walls, start, h, w := parseInput(puzzleInput)
	visited := part1(walls, start, h, w)
	fmt.Println(len(visited))
	fmt.Println(part2(visited, walls, start, h, w))
}

func parseInput(puzzleInput []string) (map[[2]int]struct{}, [2]int, int, int) {
	start := [2]int{0, 0}
	walls := make(map[[2]int]struct{})
	for i, row := range puzzleInput {
		for j := range row {
			if row[j] == '^' {
				start = [2]int{i, j}
			} else if row[j] == '#' {
				walls[[2]int{i, j}] = struct{}{}
			}
		}
	}
	h := len(puzzleInput)
	w := len(puzzleInput[0])
	return walls, start, h, w
}

func part1(walls map[[2]int]struct{}, start [2]int, h, w int) map[[2]int]struct{} {
	visited := make(map[[2]int]struct{})
	pos := start
	direction := [2]int{-1, 0}
	for {
		visited[pos] = struct{}{}
		pos2 := [2]int{pos[0] + direction[0], pos[1] + direction[1]}
		if pos2[0] < 0 || pos2[1] < 0 || pos2[0] > h-1 || pos2[1] > w-1 {
			break
		}
		if _, ok := walls[pos2]; ok {
			direction[0], direction[1] = direction[1], -direction[0]
		} else {
			pos = pos2
		}
	}
	return visited
}

func isStuckInALoop(walls map[[2]int]struct{}, start [2]int, h, w int) bool {
	visited := make(map[[4]int]struct{})
	pos := start
	direction := [2]int{-1, 0}
	for {
		state := [4]int{pos[0], pos[1], direction[0], direction[1]}
		if _, ok := visited[state]; ok {
			return true
		}
		visited[state] = struct{}{}
		pos2 := [2]int{pos[0] + direction[0], pos[1] + direction[1]}
		if pos2[0] < 0 || pos2[1] < 0 || pos2[0] > h-1 || pos2[1] > w-1 {
			break
		}
		if _, ok := walls[pos2]; ok {
			direction[0], direction[1] = direction[1], -direction[0]
		} else {
			pos = pos2
		}
	}
	return false
}

func part2(visited, walls map[[2]int]struct{}, start [2]int, h, w int) int {
	res := 0
	for pos := range visited {
		if pos == start {
			continue
		}
		walls[pos] = struct{}{}
		if isStuckInALoop(walls, start, h, w) {
			res++
		}
		delete(walls, pos)
	}
	return res
}
